### PR TITLE
Implement undo/redo for wxTextCtrl in wxOSX

### PR DESCRIPTION
**Pull request by vadz**
This is extracted from #1531 and broken in more reasonable commits, without any other changes so far.

### CODE REVIEW

File: src/osx/cocoa/textctrl.mm [hunk 1]
```diff
@@ -773,6 +773,12 @@ - (void)controlTextDidEndEditing:(NSNotification *)aNotification
 
     [tv setDelegate: tv];
 
+    m_undoManager = nil;
+    if ( wxPeer && wxPeer->GetParent() && wxPeer->GetParent()->GetPeer() && wxPeer->GetParent()->GetPeer()->GetWXWidget() )
```
Comment (vadz) [inline src/osx/cocoa/textctrl.mm:777]: @Dan-Korn I wonder if we want to use the parent here or the TLW parent? I.e. shouldn't this use `wxGetTopLevelParent()`?

File: src/osx/textentry_osx.cpp [hunk 2]
```diff
@@ -175,7 +175,7 @@ void wxTextEntry::SetInsertionPointEnd()
 
 long wxTextEntry::GetInsertionPoint() const
 {
-    long begin, end ;
+    long begin = 0, end = 0;
```
Comment (vadz) [inline src/osx/textentry_osx.cpp:178]: @Dan-Korn Do you remember what problem did this fix? It looks like `GetSelection()` always fills in the provided parameters and there are still a few places where it's called with them being uninitialized. IMO we should either initialize them everywhere, if really necessary, or drop this commit if it isn't.

**Comment by vadz** [comment]
Running the test under Mac I still get an exception when calling `Undo()` with the following stack:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = hit Objective-C exception
  * frame #0: 0x00000001966f906c libobjc.A.dylib`objc_exception_throw
    frame #1: 0x00000001976c3f5c Foundation`-[NSString substringWithRange:] + 252
    frame #2: 0x0000000199367798 AppKit`-[NSTextStorage(NSUndo) _undoRedoAttributedSubstringFromRange:] + 180
    frame #3: 0x000000019991bfa8 AppKit`-[NSUndoTyping undoRedo:] + 160
    frame #4: 0x00000001977d1d30 Foundation`-[_NSUndoStack popAndInvoke] + 188
    frame #5: 0x00000001977d1a60 Foundation`-[NSUndoManager undoNestedGroup] + 292
    frame #6: 0x0000000103233d68 libwx_osx_cocoau_core-3.1.6.0.0.dylib`wxNSTextViewControl::Undo(this=0x000000010a25bcc0) at textctrl.mm:1041:5
    frame #7: 0x0000000103053144 libwx_osx_cocoau_core-3.1.6.0.0.dylib`wxTextEntry::Undo(this=0x000000010983b880) at textentry_osx.cpp:265:24
    frame #8: 0x00000001007f49b0 test_gui`TextEntryTestCase::UndoRedo(this=0x000000016fdfa3a0) at textentrytest.cpp:368:12
    ...
```

Any idea about why does it crash?

**Comment by csomor** [comment]
> Running the test under Mac I still get an exception when calling `Undo()` with the following stack:
> Any idea about why does it crash?

yes, because there are still pointers to previous objects in the undo chain when testing, that's the thing I wrote about to modify the tests, or to determine wether we are running in the test suite, to setup things in isolated mode, I'll have to dig that out ...

**Comment by vadz** [comment]
Should we call (removeAllActionsWithTarget:)[https://developer.apple.com/documentation/foundation/nsundomanager/1409896-removeallactionswithtarget] in `wxTextCtrl` dtor to ensure no dangling operations are left?

Or am I totally missing what is this about? I admit I don't understand at all why should anything we had done previously affect undo for a freshly created control...

**Comment by csomor** [comment]
> Should we call [removeAllActionsWithTarget:](https://developer.apple.com/documentation/foundation/nsundomanager/1409896-removeallactionswithtarget) in `wxTextCtrl` dtor to ensure no dangling operations are left?

we could do this, if this is enough, then perfect, otherwise we might have to use my nuclear option ... `removeAllActions`

> Or am I totally missing what is this about? I admit I don't understand at all why should anything we had done previously affect undo for a freshly created control...

because the undo manager is shared and its references are not handled automatically

https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/UndoArchitecture/Articles/CleaningUndoStack.html

**Comment by vadz** [comment]
FWIW I do have the next commit, cherry picked from the other PR and adding `EmptyUndoBuffer()`, which wraps `removeAllActions:`. But I still have no idea where should we call it from?

Anyhow, I'll try `removeAllActionsWithTarget:` first.

**Comment by vadz** [comment]
P.S. Do I understand correctly that the undo manager is associated with the TLW, i.e. that we should use `wxGetTopLevelParent()` instead of just the immediate parent? I can't find where is it actually created, is there a built-in undo manager in Cocoa?

**Comment by csomor** [comment]
I also don't know what target is posted to the undo manager, maybe its not the control itself, but some other instance. The other option that I tried was using a private undo manager, but this also should not be necessary in real-life situations ...

**Comment by csomor** [comment]
> P.S. Do I understand correctly that the undo manager is associated with the TLW, i.e. that we should use `wxGetTopLevelParent()` instead of just the immediate parent? I can't find where is it actually created, is there a built-in undo manager in Cocoa?

yes, the NSWindow has its own, built-in. 

I'd call just before running the test

```
wxFrame* topFrame = wxDynamicCast(wxTheApp->GetTopWindow(), wxFrame);
wxOSXRemoveAllUndoActions(topFrame->GetWXWindow());
```

**Comment by vadz** [comment]
But the problem is not making the test pass. AFAICS we're going to have the same problem in the actual application code too.

**Comment by csomor** [comment]
> But the problem is not making the test pass. AFAICS we're going to have the same problem in the actual application code too.

I didn't have any crashes when testing with the samples, so I really thought we wouldn't run into this, but I fear you are right, just came across a report from someone who ran into the same issues in his app https://www.markusbodner.com/til/2021/04/30/prevent-nstextview-from-polluting-undo-history/

**Comment by vadz** [comment]
Unfortunately it looks like you were right too, calling
```
m_undoManager removeAllActionsWithTarget: m_textView];
```
in `wxNSTextViewControl` dtor isn't enough. Calling `removeAllActions` is, so I guess I'll do it, even if it's a bit problematic to lose undo history in all text controls just because one of them is destroyed -- but I don't know how to find the right target to remove. But if you have any other ideas about the target could be, I could test them...

**Comment by csomor** [comment]
> Unfortunately it looks like you were right too, calling
> 
> ```
> [m_undoManager removeAllActionsWithTarget: m_textView]
> ```
> 
> in `wxNSTextViewControl` dtor isn't enough. Calling `removeAllActions` is, so I guess I'll do it, even if it's a bit problematic to lose undo history in all text controls just because one of them is destroyed -- but I don't know how to find the right target to remove. But if you have any other ideas about the target could be, I could test them...

I'll try to implement the private undo manager

**Comment by vadz** [comment]
Sorry, I didn't want to create extra work for you... My entire goal of looking at this was to free you from having to deal with it but, of course, it didn't work, as always whenever I try to do anything with Mac :-(

But thanks in advance, it does seem that private (per-`wxTextCtrl`, I guess?) undo manager is much more compatible with wx API, so this should definitely work better. If you do this, we should just close this PR, maybe only cherry-picking 53eaf62b60 from it.

**Comment by csomor** [comment]
@vadz I've added a first attempt at a private undo manager, in my interactive tests it works, I'd be grateful if you could run the tests and tell me whether things crash somewhere ...

**Comment by csomor** [comment]
note that this only adds a private NSUndoManager and wires it back in the delegate method, so the rest of the code still remains as it is

**Comment by vadz** [comment]
Thanks, it works, even though I don't understand how -- where do we actually register this undo manager for the use in the text?

Anyhow, I'm fine with merging it as is, do you think we still want to add `EmptyUndoBuffer()`?

**Comment by csomor** [comment]
> Thanks, it works, even though I don't understand how -- where do we actually register this undo manager for the use in the text?

the delegate method `- (nullable NSUndoManager *)undoManagerForTextView:(NSTextView *)view` is responsible for making the undoManager known to the system

> Anyhow, I'm fine with merging it as is, do you think we still want to add `EmptyUndoBuffer()`?

Yes, if we have it on other platforms, then this is fine for me

**Comment by vadz** [comment]
> > Thanks, it works, even though I don't understand how -- where do we actually register this undo manager for the use in the text?
> 
> the delegate method `- (nullable NSUndoManager *)undoManagerForTextView:(NSTextView *)view` is responsible for making the undoManager known to the system

This is what I thought but somehow removing them didn't affect the test -- i.e. it still passed. So either there is something wrong with our test or I'm still missing something here.
 
> > Anyhow, I'm fine with merging it as is, do you think we still want to add `EmptyUndoBuffer()`?
> 
> Yes, if we have it on other platforms, then this is fine for me

It can be implemented for `wxTE_RICH2` controls under MSW. GTK doesn't seem to have any support for undo at all...

**Comment by csomor** [comment]
> > > Thanks, it works, even though I don't understand how -- where do we actually register this undo manager for the use in the text?
> > 
> > 
> > the delegate method `- (nullable NSUndoManager *)undoManagerForTextView:(NSTextView *)view` is responsible for making the undoManager known to the system
> 
> This is what I thought but somehow removing them didn't affect the test -- i.e. it still passed. So either there is something wrong with our test or I'm still missing something here.

no, for our tests we don't need our undo manager be known to the system, we create it, we call its instance directly and we dispose of it when our control is deleted, because of the strong reference we keep

but having this delegate is the way to support the native undo manager behaviour, although this is not active, because we don't have the native support being activated, I've disabled undo: and redo:  wiring to wxID_UNDO and wxID_REDO because this would mean we'd have to support it for all our non-native controls like wxSTC etc as well. So technically this method is not needed now, but IMHO it's better to play nicely ... you never know where this might be called from ..

**Review by csomor — APPROVED** [review]

**Comment by vadz** [comment]
I've removed the apparently unnecessary change to `wxTextEntry::GetInsertionPoint()` and added `EmptyUndoBuffer()` MSW implementation and unit test. If the CI builds pass, I'm going to merge it.

**Pull request merged**